src/supervisedoutputconnector.h

```diff
@@ -1229,6 +1247,108 @@ namespace dd
+	{
+	  if (i == 0)
+	    {
+	      cumsum.push_back(sort_pairs[i].second);
+	    }
+	  else
+	    {
+	      cumsum.push_back(cumsum.back() + sort_pairs[i].second);
+	    }
+	}
+    }
+    
+    static double compute_ap(const std::vector<std::pair<double,int>> &tp,
+			     const std::vector<std::pair<double,int>> &fp,
+			     const int &num_pos)
+    {
+      const double eps = 1e-6;
+      const int num = tp.size();
+      if (num == 0 || num_pos == 0)
+	return 0.0;
+      double ap = 0.0;
+      std::vector<int> tp_cumsum;
+      std::vector<int> fp_cumsum;
+      cumsum_pair(tp, tp_cumsum); // tp cumsum
+      cumsum_pair(fp, fp_cumsum); // fp cumsum
+      std::vector<double> prec; // precision
+      for (int i=0;i<num;i++)
+	prec.push_back(static_cast<double>(tp_cumsum[i])/(tp_cumsum[i] + fp_cumsum[i]));
+      std::vector<double> rec; // recall
+      for (int i=0;i<num;i++)
+	rec.push_back(static_cast<double>(tp_cumsum[i])/num_pos);
+
+      // voc12, ilsvrc style ap
+      float cur_rec = rec.back();
+      float cur_prec = prec.back();
+      for (int i=num-2; i>=0; --i)
+	{
+	  cur_prec = std::max<float>(prec[i], cur_prec);
+	  if (fabs(cur_rec - rec[i]) > eps)
+	    {
+	      ap += cur_prec * fabs(cur_rec - rec[i]);
+	    }
+	  cur_rec = rec[i];
+	}
+      ap += cur_rec * cur_prec;
+      return ap;
+    }
+    
+    static double ap(const APIData &ad)
+    {
+      double mAP = 0.0;
+      std::map<int, float> APs;
+      
+      // extract tp, fp, labels
+      APIData bad = ad.getobj("0");
+      int pos_count = ad.get("pos_count").get<int>();
+      for (int i=0;i<pos_count;i++)
+	{
+	  std::vector<APIData> vbad = bad.getv(std::to_string(i));
+	  //std::cerr << "vbad size=" << vbad.size() << std::endl;
+	  for (size_t j=0;j<vbad.size();j++)
+	    {
+	      std::vector<double> tp_d = vbad.at(j).get("tp_d").get<std::vector<double>>();
+	      //std::cerr << "tp_d size=" << tp_d.size() << std::endl;
+	      std::vector<int> tp_i = vbad.at(j).get("tp_i").get<std::vector<int>>();
+	      std::vector<double> fp_d = vbad.at(j).get("fp_d").get<std::vector<double>>();
+	      std::vector<int> fp_i = vbad.at(j).get("fp_i").get<std::vector<int>>();
+	      int num_pos = vbad.at(j).get("num_pos").get<int>();
+	      //std::cerr << "num_pos=" << num_pos << std::endl;
+	      int label = vbad.at(j).get("label").get<int>();
+	      //std::cerr << "label=" << label << std::endl;
+	      std::vector<std::pair<double,int>> tp;
+	      std::vector<std::pair<double,int>> fp;
+	      
+	      //std::cerr << "fp_d size=" << fp_d.size() << std::endl;
+	      for (size_t j=0;j<tp_d.size();j++)
+		{
+		  tp.push_back(std::pair<double,int>(tp_d.at(j),tp_i.at(j)));
+		}
+	      for (size_t j=0;j<fp_d.size();j++)
+		{
+		  fp.push_back(std::pair<double,int>(fp_d.at(j),fp_i.at(j)));
+		}
+	      
+	      APs[label] = compute_ap(tp,fp,num_pos);
+	      //std::cerr << "ap for label " << label << "=" << APs[label] << std::endl;
+	      mAP += APs[label];
```

Comment:
ici ça agrège pas comme il faut: "i" itère sur les images, et "j" sur les labels présents sur une image
or dans ce code ça somme le AP de chaque label au résultat global et ensuite ça divise en dessous le résultat global par le nombre de classes de l'image courante, du coup les premiers scores sont divisés bcp trop souvent